dynoid/dynoid.go

```diff
@@ -5,43 +5,230 @@ import (
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
+	"log/slog"
+	"os"
+	"path"
 	"strings"
 	"sync"
 
 	"github.com/coreos/go-oidc/v3/oidc"
 )
 
-type IssuerCallback func(string) error
+const (
+	defaultAudience = "heroku"
 
-type Verifier struct {
-	IssuerCallback IssuerCallback
-	config         *oidc.Config
-	providers      sync.Map
+	//nolint: gosec
+	defaultTokenPath = "/etc/heroku/dyno_id_token"
+)
+
+// Returned by an IssuerCallback getting an issuer it doesn't trust
+type UntrustedIssuerError struct {
+	Issuer string
 }
 
-func New(clientID string) *Verifier {
-	return &Verifier{config: &oidc.Config{ClientID: clientID}}
+func (e *UntrustedIssuerError) Error() string {
+	return fmt.Sprintf("untrusted issuer: %v", e.Issuer)
+}
+
+// Returned when the token doesn't match the expected format
+type MalformedTokenError struct {
+	Err error
+}
+
+func (e *MalformedTokenError) Error() string {
+	return fmt.Sprintf("malformed token: %s", e.Err.Error())
 }
 
+func (e *MalformedTokenError) Unwrap() error {
+	return e.Err
+}
+
+type staticError string
+
+func (e staticError) Error() string {
+	return string(e)
+}
+
+const (
+	ErrMustCheckIssuer staticError = "must check issuer"
+)
+
+// An IssuerCallback is called whenever a token is verified to ensure it matches
+// some expected criteria.
+type IssuerCallback func(issuer string) error
+
+// AllowHerokuHost verifies that the issuer is from Heroku for the given host
+// domain
 func AllowHerokuHost(host string) IssuerCallback {
 	return func(issuer string) error {
 		if !strings.HasPrefix(issuer, fmt.Sprintf("https://oidc.%v/", host)) {
-			return fmt.Errorf("untrusted issuer: %v", issuer)
+			return &UntrustedIssuerError{Issuer: issuer}
 		}
 
 		return nil
 	}
 }
 
+// Subject contains all of the subject information stored by Heroku when issuing
+// a token.
+type Subject struct {
+	AppID   string `json:"app_id"`
+	AppName string `json:"app_name"`
+	Dyno    string `json:"dyno"`
+}
+
+func (s *Subject) LogValue() slog.Value {
+	return slog.GroupValue(
+		slog.String("app_id", s.AppID),
+		slog.String("app_name", s.AppName),
+		slog.String("dyno", s.Dyno),
+	)
+}
+
+func (s *Subject) MarshalText() ([]byte, error) {
+	return []byte(s.String()), nil
+}
+
+func (s *Subject) UnmarshalText(text []byte) error {
+	if s == nil {
+		*s = Subject{}
+	}
+
+	sub := string(text)
+	parts := strings.Split(sub, ":")
+	if len(parts) != 5 || parts[0] != "app" || parts[3] != "dyno" {
+		return fmt.Errorf("unexpected subject format: %q", sub)
+	}
+
+	app := strings.Split(parts[1], ".")
+	if len(app) != 2 {
+		return fmt.Errorf("unexpected subject format: %q", sub)
+	}
+
+	s.AppID = app[0]
+	s.AppName = app[1]
+	s.Dyno = parts[4]
+
+	return nil
+}
+
+func (s *Subject) String() string {
+	if s == nil {
+		return ""
+	}
+
+	return fmt.Sprintf("app:%s.%s::dyno:%s", s.AppID, s.AppName, s.Dyno)
+}
+
+// Subject contains all of the token information stored by Heroku when issuing
```

Comment:
Did you intend to rewrite this comment to Token specifically, or remove it?

Reply:
Good catch! Fixed.